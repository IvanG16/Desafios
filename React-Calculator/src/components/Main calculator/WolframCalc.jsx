import React, { useState } from 'react';
import CalculatorShell from '../CalculatorShell.jsx';
import '../Calculator.css';
import { operators, determinesOutput } from './utils/index.js';

export function WolframCalc({ setLogList }) {
    const [isOff, setIsOff] = useState(false);
    const [number1, setNumber1] = useState(null);
    const [operator, setOperator] = useState(null);
    const [accumulator, setAccumulator] = useState('');

    async function addData (input) {
        if(isOff && input !== 'Off'){
            return;
        }

        if (operators.includes(input) && !number1) {
                setNumber1(accumulator);
                setAccumulator('');
                setOperator(input);
                return;
        }

        determinesOutput({ input, setAccumulator, accumulator, setNumber1, number1, setOperator, operator, setIsOff, isOff ,setLogList })
    }

    return <CalculatorShell accumulator={accumulator} isOff={isOff} addData={addData} type={'wolfram'}/>;
}

export default WolframCalc;
