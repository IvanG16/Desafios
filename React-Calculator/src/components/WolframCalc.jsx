import React, { useState } from 'react';
import CalculatorShell from './CalculatorShell';
import './Calculator.css';
import { number } from 'prop-types';

export const operations = {
    'âˆš': (x) => parseFloat(x) + parseFloat(x),
    '^': (x, y) => parseFloat(x) - parseFloat(y),
    'cos': (x, y) => parseFloat(x) * parseFloat(y),
    'tan': (x, y) => parseFloat(x) / parseFloat(y),
    'sen': (x, y) => parseFloat(y) * (parseFloat(x) / 100),
};
export const operators = Object.keys(operations);

const getApiData = async (number1, operator, accumulator) => {
    if(!number1 && operator != '^'){
        const response = await fetch(`/api/v2/query?appid=XA3Y4W-2EP6LQJ3W6&input=${operator}%20${accumulator}&output=json`);
        if (response.ok) {
            const data = await response.json();
            if (data.queryresult.success) {
                const apiResponse = (data.queryresult.pods.find(({ id }) => id === 'Result')) ? data.queryresult.pods.find(({ id }) => id === 'Result')?.subpods[0]?.plaintext : data.queryresult.pods.find(({ id }) => id === 'DecimalApproximation')?.subpods[0]?.plaintext
                return apiResponse.substring(0, 28);
            }
        }
    }else if(number1){
        return 'Syntax Error';
    }
    throw new Error(`Error al calcular ${number1} ${operator} ${accumulator}`);
};

async function inputtedEquals({ number1, operator, accumulator, setAccumulator, setLogList }){
    if(!operator){
        return console.log("Nothing to operate")
    }

    const apiData = await getApiData(number1, operator, accumulator);
    setAccumulator(apiData);
    setLogList((previousLogs) => {
        const newLogs = {
            calculo: `${number1} ${operator} ${accumulator}`,
            date: new Date(),
        }
        return [...previousLogs, newLogs]
    }
    )
}

function inputtedClearAll({ setAccumulator, setNumber1, setOperator }){
    setAccumulator('');
    setNumber1(null);
    setOperator(null);
}

function inputtedDelete({ setAccumulator, accumulator, setOperator, operator, setNumber1, number1 }){
    if (accumulator) return setAccumulator(accumulator.substring(0, accumulator.length - 1));
    
    if (operator) return setOperator(operator.substring(0, operator.length - 1));

    if (number1) return setNumber1(number1.substring(0, number1.length - 1));
}

function onOff({ isOff, setIsOff, setAccumulator, setOperator, setNumber1 }){
    setIsOff(!isOff);
    setAccumulator('');
    setNumber1(null);
    setOperator(null);
}

function inputtedDecimalPoint({ setAccumulator, accumulator, input }){
    if (accumulator.includes('.')) return setAccumulator(accumulator);

    setAccumulator(accumulator + input);
}

function inputtedNumber({ accumulator, setAccumulator, input }){
    if(accumulator.length < 28) return setAccumulator(accumulator + input);
}

function refactoredSwitch({ input, setAccumulator, accumulator, setNumber1, number1, setOperator, operator, setIsOff, isOff ,setLogList }){
    if(inputs[input]){
        return inputs[input]({ input, setAccumulator, accumulator, setNumber1, number1, setOperator, operator, setIsOff, isOff ,setLogList });
    }
    return inputtedNumber({ accumulator, setAccumulator, input });
}

const inputs = {
    ['=']: inputtedEquals,
    ['C']: inputtedClearAll,
    ['del']: inputtedDelete,
    ['Off']: onOff,
    ['.']: inputtedDecimalPoint,
    ...operations,
}

export function WolframCalc({ setLogList }) {
    const [isOff, setIsOff] = useState(false);
    const [number1, setNumber1] = useState(null);
    const [operator, setOperator] = useState(null);
    const [accumulator, setAccumulator] = useState('');

    async function addData (input) {
        if(isOff && input !== 'Off'){
            return;
        }

        if (operators.includes(input) && !number1) {
                setNumber1(accumulator);
                setAccumulator('');
                setOperator(input);
                return;
        }

        refactoredSwitch({ input, setAccumulator, accumulator, setNumber1, number1, setOperator, operator, setIsOff, isOff ,setLogList })
    }

    return <CalculatorShell accumulator={accumulator} isOff={isOff} addData={addData}/>;
}

export default WolframCalc;
