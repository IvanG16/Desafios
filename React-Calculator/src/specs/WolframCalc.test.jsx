import WolframCalc from '../components/WolframCalc'
import { describe, it, expect, afterEach, vi} from 'vitest'
import { fireEvent, render, cleanup } from '@testing-library/react'

/**
 * @vitest-environment jsdom
 */

describe(' <WolframCalc />', ()=> {
    const handleOnClick = vi.fn();

    afterEach(cleanup);

    it('renders properly', ()=> {
        const { container } = render(<WolframCalc />);
        expect(container).toMatchSnapshot()
    })
    
})
