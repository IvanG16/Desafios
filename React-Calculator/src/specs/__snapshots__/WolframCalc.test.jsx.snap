// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[` <WolframCalc /> > renders properly 1`] = `
<div>
  <div
    class="w-[425px] h-[520px] bg-orange-500 rounded-[20px] flex-col justify-start items-start inline-flex"
  >
    <div
      class="w-[425px] h-[147px] px-2.5 py-5 rounded-[20px] border border-neutral-500 flex-col justify-start items-end gap-2.5 inline-flex"
    >
      <div
        class="w-[134px] h-[27px] bg-stone-600"
      />
      <div
        class="w-[401px] h-[70px] justify-end items-center gap-2.5 inline-flex bg-stone-600"
      >
        <div
          class="w-[82px] h-[60px] px-4 text-white text-3xl font-normal font-['radioland'] flex justify-end items-center tracking-wide"
        />
      </div>
    </div>
    <div
      class="w-[425px] h-[373px] px-2.5 py-[30px] justify-start items-start gap-[25px] inline-flex items-center"
    >
      <div
        class="justify-start items-start gap-[25px] inline-flex flex-wrap items-center"
      >
        <div
          class="justify-start items-start gap-[25px] inline-flex items-center"
        >
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-50 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              Off
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-50 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              C
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-50 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              del
            </div>
          </button>
        </div>
        <div
          class="justify-start items-start gap-[25px] inline-flex flex-wrap items-center"
        >
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              7
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              8
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              9
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              4
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              5
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              6
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              1
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              2
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              3
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              .
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              0
            </div>
          </button>
          <button
            class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-300 "
            type="button"
          >
            <div
              class="w-1/2 text-black text-2xl font-normal Inconsolata"
            >
              =
            </div>
          </button>
        </div>
      </div>
      <div
        class="justify-start items-start gap-[25px] inline-flex flex-wrap items-center"
      >
        <button
          class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-50 "
          type="button"
        >
          <div
            class="w-1/2 text-black text-2xl font-normal Inconsolata"
          >
            âˆš
          </div>
        </button>
        <button
          class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-50 "
          type="button"
        >
          <div
            class="w-1/2 text-black text-2xl font-normal Inconsolata"
          >
            ^
          </div>
        </button>
        <button
          class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-50 "
          type="button"
        >
          <div
            class="w-1/2 text-black text-2xl font-normal Inconsolata"
          >
            cos
          </div>
        </button>
        <button
          class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-50 "
          type="button"
        >
          <div
            class="w-1/2 text-black text-2xl font-normal Inconsolata"
          >
            tan
          </div>
        </button>
        <button
          class="w-[81px] h-[45px] p-2.5 rounded-[5px] border border-black flex-col justify-center items-center gap-2.5 inline-flex bg-stone-50 "
          type="button"
        >
          <div
            class="w-1/2 text-black text-2xl font-normal Inconsolata"
          >
            sen
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
